pkg/lib/dialogs.tsx

```diff
@@ -153,41 +158,28 @@ export const useDialogs = () => {
 };
 
 export const WithDialogs = ({ children } : { children: React.ReactNode }) => {
-    const is_open = useRef(false); // synchronous
-    const resolveRef = useRef<((x: unknown) => void) | null>(null);
-    const rejectRef = useRef<((x: unknown) => void) | null>(null);
     const [dialog, setDialog] = useState<React.ReactNode>(null);
+    const shown = useRef<"show" | "run" | "close">("close");
+
+    function transition(expected: string, to: typeof shown.current, arg: React.ReactNode = null) {
+        if (shown.current !== expected)
+            throw new Error(`Dialogs.${to}(${JSON.stringify(arg)}) called, but that's only valid` +
+                            `after .${expected}(), current dialog is ${JSON.stringify(dialog)}.`);
+        shown.current = to;
+        setDialog(arg);
+    }
 
     const Dialogs: Dialogs = {
-        show: component => {
-            if (component && is_open.current)
-                console.error("Dialogs.show() called for",
-                              JSON.stringify(component),
-                              "while a dialog is already open:",
-                              JSON.stringify(dialog));
-            is_open.current = !!component;
-            setDialog(component);
-            return new Promise((resolve, reject) => {
-                resolveRef.current = resolve;
-                rejectRef.current = reject;
-            });
-        },
-        close: (args) => {
-            is_open.current = false;
-            setDialog(null);
-            if (resolveRef.current !== null) {
-                resolveRef.current(args);
-                resolveRef.current = null;
-                rejectRef.current = null;
-            }
-        },
-        reject: (err) => {
-            is_open.current = false;
-            setDialog(null);
-            if (rejectRef.current !== null) {
-                rejectRef.current(err);
-                resolveRef.current = null;
-                rejectRef.current = null;
+        show: (component: React.ReactNode) => transition("close", "show", component),
+        close: () => transition("show", "close"),
+        run: async (component, properties) => {
+            try {
+                return await new Promise((resolve, reject) => {
+                    const dialog = React.createElement(component, { ...properties, result: { resolve, reject } });
```

Comment:
FTR, `{...undefined }` works fine, so not specifying `properties` at all is legit.

Reply:
See the comment above — this isn't about runtime behaviour, but about type safety and inference.

Reply:
So TypeScript doesn't understand the spread operator? `{...undefined}` should match any dictionary type.

Reply:
This would be fine as well — the issue is with the type of the function.  It's

```ts
run<T, P>(
    component: React.ComponentType<P & { dialogResult: DialogResult<T> }>,
    properties: P
): Promise<T>;
```

We could consider a few approaches but none of them seem to work well:
 - simply making the properties argument optional would mean that you could skip properties for all dialogs, not just ones that actually take no properties.  TypeScript prevents us from doing this because it makes the properties array we pass into `createElement()` no longer match the one required by the component type.
 - we could maybe figure out a way to define a function that works with overloaded definitions which take different numbers of arguments, but I can't figure out how to do this when the `interface` is defined separately from the code.  (This separate definition of the interface was necessitated by how we have no `this` when `Dialogs.close` is used directly from a event handler.)
 - we could try some exotic type transformation like OrUndefined<P> which adds undefined as a valid value for the properties array only if `P` is the empty object, but Typescript again doesn't enjoy reasoning about this at the point of the spread operator.  I wasn't able to understand why it considers it unsound, but my experience tells me that it's probably right. :)
 - we could add a new separate function for opening dialogs without args, but that seems kinda ugly...